Nodejs/Product/TestAdapter/TestFilesUpdateWatcher.cs

```diff
@@ -1,40 +1,36 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
 using System.Collections.Generic;
-using System.IO;
+using System.Diagnostics;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell.Interop;
 using Microsoft.VisualStudio.TestPlatform.ObjectModel;
 
 namespace Microsoft.VisualStudioTools.TestAdapter
 {
-    internal class TestFilesUpdateWatcher : IDisposable
+    internal class TestFilesUpdateWatcher : IVsFileChangeEvents
     {
-        private readonly IDictionary<string, FileSystemWatcher> _fileWatchers;
+        private readonly IDictionary<string, uint> watchedFiles = new Dictionary<string, uint>(StringComparer.OrdinalIgnoreCase);
         public event EventHandler<TestFileChangedEventArgs> FileChangedEvent;
 
-        public TestFilesUpdateWatcher()
+        private readonly IVsFileChangeEx fileWatcher;
+
+        public TestFilesUpdateWatcher(IServiceProvider serviceProvider)
         {
-            _fileWatchers = new Dictionary<string, FileSystemWatcher>(StringComparer.OrdinalIgnoreCase);
+            this.fileWatcher = serviceProvider.GetService<IVsFileChangeEx>(typeof(SVsFileChangeEx));
         }
 
-        public bool AddWatch(string path)
+        public bool AddFileWatch(string path)
         {
             ValidateArg.NotNull(path, "path");
 
-            if (!string.IsNullOrEmpty(path))
+            const uint mask = (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Add | _VSFILECHANGEFLAGS.VSFILECHG_Del | _VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time);
+
+            if (!this.watchedFiles.ContainsKey(path) && ErrorHandler.Succeeded(this.fileWatcher.AdviseFileChange(path, mask, this, out uint cookie)))
             {
-                var directoryName = Path.GetDirectoryName(path);
-                var filter = Path.GetFileName(path);
-
-                if (!_fileWatchers.ContainsKey(path) && Directory.Exists(directoryName))
-                {
-                    var watcher = new FileSystemWatcher(directoryName, filter);
-                    _fileWatchers[path] = watcher;
-
-                    watcher.Changed += OnChanged;
-                    watcher.EnableRaisingEvents = true;
-                    return true;
-                }
+                this.watchedFiles.Add(path, cookie);
```

Comment:
Presumably, there's some reason to believe calls will be sequentialized that we don't have to lock `watchedFiles`?

Reply:
Yes, all of this happens on the UI thread